### PR TITLE
[Branch View] manual merge tag

### PR DESCRIPTION
**Pull request by 6543**
(https://github.com/go-gitea/gitea/issues/7976)  Show Merge Tag if it is merged manualy

# UI:
manual-merge-demo

### CODE REVIEW

**Comment by lafriks** [comment]
Imho this should be checked on push and saved to pr

**Comment by 6543** [comment]
> Imho this should be checked on push and saved to pr

I dont understand? What do you mean?

**Comment by 6543** [comment]
Do you mean this https://github.com/go-gitea/gitea/issues/7976#issuecomment-524673002 ?

**Comment by lafriks** [comment]
> > Imho this should be checked on push and saved to pr
> 
> I dont understand? What do you mean?

Nevermind, it's late I thought it is about PR :)

**Comment by 6543** [comment]
do it the short way ... https://github.com/go-gitea/gitea/issues/7976#issuecomment-524934617

**Comment by 6543** [comment]
> Nevermind, it's late I thought it is about PRs manual merge displaying :)

@lafriks but what dou you think of this solution to https://github.com/go-gitea/gitea/issues/7976 anyway

**Comment by 6543** [comment]
WIP notice: combine if statements

**Review by lunny — APPROVED** [review]

**Comment by 6543** [comment]
@lafriks can you review it?

File: options/locale/locale_en-US.ini [hunk 1]
```diff
@@ -1409,6 +1409,7 @@ branch.restore_failed = Failed to restore branch '%s'.
 branch.protected_deletion_failed = Branch '%s' is protected. It cannot be deleted.
 branch.restore = Restore Branch '%s'
 branch.download = Download Branch '%s'
+branch.manual_merged = Branch was merged manualy
```
Comment (gary-kim) [inline options/locale/locale_en-US.ini]:
```suggestion
branch.manual_merged = Branch was merged manually
```

**Review by gary-kim — CHANGES_REQUESTED** [review]
Just to ask, is this also going to indicate a branch is merged if the branch was just created and pushed from the master branch?

**Comment by 6543** [comment]
> Just to ask, is this also going to indicate a branch is merged if the branch was just created and pushed from the master branch?

It does. -> https://github.com/go-gitea/gitea/issues/7976#issuecomment-525257557

**Comment by gary-kim** [comment]
> > Just to ask, is this also going to indicate a branch is merged if the branch was just created and pushed from the master branch?
> 
> It does. -> [#7976 (comment)](https://github.com/go-gitea/gitea/issues/7976#issuecomment-525257557)

IMHO, having those branches that have no commits ahead of the master being called merged seems a bit weird as it could also mean that the branch was just created. I feel like that is going to cause some confusion.

**Comment by 6543** [comment]
we can check if it is at least one commit behind the default one to avoid souch or if it doesnt use to mouch cpu time lookup the merge commit

**Comment by lunny** [comment]
We also could detect if the second commit of the branch is not the second one of the default branch.

**Comment by 6543** [comment]
> We also could detect if the second commit of the branch is not the second one of the default branch.

nice idear this would save lookup time and works with a lot of constructs ...
but if something is merged without --no-ff and no commits betwen it will not show up as merged ...
but i personaly have no problem with that :D

**Comment by 6543** [comment]
> > We also could detect if the second commit of the branch is not the second one of the default branch.

> `current_branch.comit.parentcommit == current_branch.comit.parentcommit -> branch merged = false` ??

to test if divergence.Behind != 0 should have the same impact?

**Comment by 6543** [comment]
we could use [`git branch --merged | grep $branch`](https://stackoverflow.com/questions/226976/how-can-i-know-if-a-branch-has-been-already-merged-into-master)

**Comment by 6543** [comment]
I cleanup the 22 commits so it is easy to understand - the actual diff is very smal ...

**Comment by 6543** [comment]
# Actual State:

manual merge tag if:
 * no commits ahead default branch
 * not same as default branch

# UI:

manual-merge-demo

**Review by guillep2k — APPROVED** [review]

**Comment by 6543** [comment]
thx now its ready to merge :rocket:

**Review by gary-kim — APPROVED** [review]

File: routers/repo/branch.go [hunk 2]
```diff
@@ -203,10 +204,16 @@ func loadBranches(ctx *context.Context) []*Branch {
 			}
 		}
 
+		isMerged := true
+		if (divergence.Ahead != 0) || (divergence.Behind == 0) || (ctx.Repo.Repository.DefaultBranch == branchName) {
```
Comment (lunny) [inline routers/repo/branch.go]: How about to change this to

```
isMerged := false
if (divergence.Ahead == 0 && divergence.Behind > 0 && ctx.Repo.Repository.DefaultBranch != branchName)  {
     isMerged = true
}
```

**Review by lunny — CHANGES_REQUESTED** [review]

File: routers/repo/branch.go (hunk 2 above)
Comment (6543) [reply routers/repo/branch.go]: well the result will be the same
 + it if false by default
 -  it has to check each condition to set it to true (speed)

**Review by 6543 — COMMENTED** [review]
@lunny changed it ...

File: routers/repo/branch.go [hunk 3]
```diff
@@ -203,10 +204,16 @@ func loadBranches(ctx *context.Context) []*Branch {
 			}
 		}
 
+		isMerged := false
+		if divergence.Ahead == 0 && divergence.Behind > 0 && ctx.Repo.Repository.DefaultBranch != branchName {
```
Comment (lunny) [inline routers/repo/branch.go:208]: And I thought it again, maybe `divergence.Behind > 0` should be `divergence.Behind >= 0`? Consider a branch just merged to `master` and no new commits pushed to master.

Comment (6543) [reply routers/repo/branch.go:208]: but this also let branches who are clones of default-branch show up as "merged":

```
git checkout -b testbranch origin/master
git push -u origin testbranch
```

**Comment by 6543** [comment]
Close thiss issue because of brainstorm with lunny

```
...
lunny
So Is it possible to know it's a cloned branch or a merged branch?

6543
as far as i know there is no way cou can be sure

6543
i got an idear but this will add a fourth lable to the branch view
if we dont calle it "manualy merged" but "included"
...
```

**Pull request closed without merging**